dialogue-core/src/main/java/com/palantir/dialogue/core/StickyEndpointChannels2.java

```diff
@@ -158,12 +165,35 @@ public String toString() {
                     // the first call.
```

Comment:
With this comment was I was getting at is: ListenableFutures does not guarantee ordering of listeners. I want to make sure that after this call is complete, I first update my internal state (stickTarget etc.) BEFORE anything that's queued up on this future can inspect this internal state.

I thought that #addDirectCallback would basically not guarantee this at all, hence why I went for transform + catchingAllAsync. Maybe there's a better way to achieve it or I am misunderstanding something?

Reply:
Ok, I've read the impl a bit, and I think this should totally be a DialogueFutures.addDirectCallback, instead of this malarky: the future internally builds a linked list, in-order of the calls (so the order is deterministic if done from single thread, which we do in here.

Reply:
```
        private final FutureCallback<Response> callInFlightResponseCallback = new FutureCallback<>() {
            @Override
            public void onSuccess(Response result) {
                try {
                    successfulCall(result);
                } catch (Throwable t) {
                    result.close();
                    throw new SafeIllegalStateException("Failed to update state with successful call", t);
                }
            }

            @Override
            public void onFailure(Throwable _ignored) {
                failed();
            }
        };

// and then
result = executeWithStickyToken(request, endpointChannel);
                    callInFlight = result;
                    DialogueFutures.addDirectCallback(result, callInFlightResponseCallback);
```

Reply:
Good call, it's much simpler that way. We don't need to worry about catch/close in that case either since it's not a transformation.

Reply:
Of course I was wrong here: see internal issue PDS-241899; The future does internally build a linked list, except that when a future completes is sets the value immediately and then executes the listeners. And that's all great, but #addListener skips the queue and executes the listener immediately if this happens :(

So I'll introduce an intermediate SettableFuture here, to make sure that the callInFlightCallback executes first and updates the internal state BEFORE executing all the futures waiting on it!